tests/PySys/monitoring/monitoring_with_collectd/run.py

```diff
@@ -0,0 +1,124 @@
+from pysys.basetest import BaseTest
+
+import time
+import re
+import json
+
+"""
+Validate tedge-dm-agent  messages that are published
+on tedge/measurements
+
+Given a configured system
+When we start the collectd with sudo in the background
+When we start the tedge-dm-agent with sudo in the background
+When we start tedge sub with sudo in the background
+Wait for couple of seconds to publish couple of batch of messages
+Then we kill tedge sub with sudo as it is running with a different user account
+Then we validate the  messages in the output of tedge sub,
+"""
+
+
+class MonitoringWithCollectd(BaseTest):
+
+    def setup(self):
+        self.js_msg = ""
+        self.cpu_cnt = 0
+        self.memory_cnt = 0
+        self.time_cnt = 0
+        self.disk_cnt = 0
+
+    def execute(self):
+        tedge = "/usr/bin/tedge"
+        sudo = "/usr/bin/sudo"
+
+        collectd = self.startProcess(
+            command=sudo,
+            arguments=["systemctl", "start", "collectd"],
+            stdouterr="collectd",
+        )
+
+        collectd_mapper = self.startProcess(
+            command=sudo,
+            arguments=["systemctl", "start", "tedge-dm-agent"],
+            stdouterr="collectd_mapper",
+        )
+
+        time.sleep(0.1)
+
+        sub = self.startProcess(
+            command=sudo,
+            arguments=[tedge, "mqtt", "sub", "--no-topic", "tedge/#"],
+            stdouterr="tedge_sub",
+            background=True,
+        )
+
+        # Wait for a small amount of time to give tedge sub time
+        # to initialize and capture couple of batches of messages
+        # that are published by tedge-dm-agent.
+        time.sleep(12)
+
+        # Kill the subscriber process explicitly with sudo as PySys does
+        # not have the rights to do it
+        kill = self.startProcess(
+            command=sudo,
+            arguments=["killall", "tedge"],
+            stdouterr="kill_out",
+        )
+
+    def validate(self):
+        self.assertGrep("tedge_sub.out", r'time|cpu|memory|df-root')
+        self.assertThat('collectd_msg_validation_result == expected_result',
+                        collectd_msg_validation_result=self.validate_json(), expected_result=True)
```

Comment:
Thinking more about the way this validation is done here, this will make the debugging difficult when there's a failure. We will have no clue why the test failed, whether it's because of a missing `cpu` field, `time` field or `df` field. This will just tell us that the validation failed and then we'll have to dig deeper into the output files to see where the failure happened. Instead it would be better if the assertion is done for each individual field in the methods below so that the assertion will clearly highlight which validation failed.

Reply:
I agree, adding assert for every parameter validation will be too many messages. So, I did go for this style, now I plan to print the reason why it failed. This will give us more info about the failure as well as we can avoid all the assert calls.
For example, something like below will help us to get the reason if the validation of a particular parameter fails, here it's CPU validation.

reason = "cpu stat validation failed in message: " + str(line)
self.abort(False, reason)